crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -5554,7 +5560,48 @@ impl<'db> TypeInferenceBuilder<'db> {
                         }
                     }
                 }
-                bindings.return_type(self.db())
+
+                let db = self.db();
+                let scope = self.scope();
+                let return_ty = bindings.return_type(db);
+
+                let find_narrowed_symbol = || match arguments.args.first() {
+                    None => {
+                        // This branch looks extraneous, especially in the face of
+                        // `missing-arguments`. However, that lint won't be able to catch this:
+                        //
+                        // ```python
+                        // def f(v: object = object()) -> TypeIs[int]: ...
+                        //
+                        // if f(): ...
+                        // ```
+                        if let Some(builder) = self
+                            .context
+                            .report_lint(&INVALID_TYPE_GUARD_CALL, arguments)
+                        {
+                            builder.into_diagnostic("Type guard call does not have a target");
+                        }
+                        None
+                    }
+                    Some(ast::Expr::Name(ast::ExprName { id, .. })) => {
+                        let name = id.as_str();
+                        let symbol = symbol_table(db, scope).symbol_id_by_name(name)?;
+
+                        Some((symbol, name.to_string()))
+                    }
+                    // TODO: Attribute and subscript narrowing
+                    Some(_) => None,
+                };
+
+                // TODO: Handle unions/intersections
```

Comment:
What kind of handling would we expect here? I don't think it's valid to expect narrowing from returning a union where one of the elements is a `TypeIs`...